DC-935, DC-932, DC-930: Update render configs with new arguments; Update Getting Started Docs

snf2ye opened the pull request:
https://broadworkbench.atlassian.net/browse/DC-935
https://broadworkbench.atlassian.net/browse/DC-932
https://broadworkbench.atlassian.net/browse/DC-930
______________
## TDR Developer take note: You may want to make these changes in your local setup
Note: None of these changes are breaking changes. These variables were previously not mentioned in our getting started instruction nor in our render configs script. If you continue on without making changes, nothing will be impacted. However, you won't be able to get full use of the new flags on the render-configs script.

### Update your environment variables
**Add** the following variables (these are now populated in render-configs based on defaults/flags)
* export AZURE_SYNAPSE_WORKSPACENAME=$(cat /tmp/azure-synapse-workspacename.txt)
* export RBS_POOLID=$(cat /tmp/rbs-pool-id.txt)
* export RBS_INSTANCEURL=$(cat /tmp/rbs-instance-url.txt)

**Remove the Azure B2C settings from env variables** - They're no longer needed since I've moved them into app.properties.

## Included Changes in PR
### Update render configs
- If you don't add any arguments, the ./render-configs.sh script should run the same as before
- There are 3 new flags:
   (1) `-a`: Azure - Options: dev, integration - This sets the azure environment for synapse and the azure tenant. Users should use dev by default and switch to integration when running Azure integration tests.
  (2) `-r`: RBS - Options: tools, dev - This sets the RBS environment. If you're just running tests locally, you should use tools. The GCP projects created locally will get deleted in about 24 hours after creation.  If you want a longer lived project, you should use the dev environment. 
  (3) `-i`: Intellij - This flag prints puts the environment variables and their values into your clipboard so that they can be easily copied into an intellij test/run profile.
- We now write to three new temporary files (noted above) to facilitate the switch in environments for Azure and RBS. 
- It's still a good option to maintain your Z-shell configuration (.zshrc) with the environment variables pointing to the tmp files where we write our the keys/environment-specific values out to. 

### Updated Getting Started Instructions
These updates should reflect the latest best practices.
- Don't automatically set up a personal dev env for onboarding. Moved these instructions to an appendix.
- More detailed instructions for running tests
- Removed some of the deployment information on the main readme and pointed to the terra support docs. 

### Updated Setup Scripts Readme
There should now be instructions for settings up OMOP datasets on Azure and GCP which is needed for testing snapshot builder settings.

snf2ye committed:
Update render configs to take an env argument and set azure env variables

snf2ye committed:
More specific variable naming

snf2ye committed:
remove token

pshapiro4broad commented on render-configs.sh:
```diff
@@ -1,32 +1,43 @@
 
-vault read -field=tenant-id secret/dsde/datarepo/dev/azure-application-secrets \
-    > /tmp/jade-dev-tenant-id.key
+vault read -field=tenant-id secret/dsde/datarepo/${AZURE_ENV}/azure-application-secrets \
```
Could you run `shellcheck` on this script? I think these env vars should have double quotes around them.
```suggestion
vault read -field=tenant-id secret/dsde/datarepo/"${AZURE_ENV}"/azure-application-secrets \
```

snf2ye committed:
shell check and reorg

snf2ye committed:
[Rough draft] Add RBS settings as well

snf2ye committed:
wrong workspace

snf2ye committed:
Remove the instructions to terraform another data repo - right now our terraform is not in a good state and we plan on moving it soon

snf2ye committed:
Move to getting-started-docs

snf2ye committed:
wip

snf2ye committed:
print settings

snf2ye committed:
wip

snf2ye committed:
Add azure managed app instrutions

snf2ye committed:
Add test instructions

snf2ye committed:
Move bash flags to options

snf2ye committed:
Update the flags in the docs

snf2ye commented on docs/jade-getting-started.md:
```diff
@@ -199,35 +205,7 @@ git clone git@github.com:broadinstitute/terraform-ap-deployments.git
-
-## 9. Google Cloud Platform setup
-
-Throughout these instructions, replace all instances of `ZZ` with your initials.
```
I've moved these steps to the bottom of the setup instructions since we're trying to move away from everyone needing to set up a dev environment.

snf2ye commented on docs/jade-getting-started.md:
```diff
@@ -311,30 +247,170 @@ psql -f db/create-data-repo-db
+
+## 14. Running locally with other locally running services
+1. Sam - set environment variable `SAM_BASEPATH` to `https://local.broadinstitute.org:50443`
+
```
I moved this from the main README. We could also get remove it altogether.

snf2ye committed:
Switch back to previous file paths so this doesn't break things for other developers

snf2ye committed:
add quote

snf2ye committed:
unset vars

snf2ye committed:
Add instruction

rjohanek reviewed (approved):
Thanks Shelby!

snf2ye committed:
Moving params from getting-started docs to app.properties

Follow-on to this: https://github.com/DataBiosphere/jade-data-repo/pull/1624/files

Remaining question: How is the client secret set? This wasn't in the getting started instructions

snf2ye committed:
Revert "Moving params from getting-started docs to app.properties"

This reverts commit 487182ebf0b1c053ec3f559a26df37a697e35c07.

snf2ye committed:
Add back instructions

snf2ye committed:
Add OMOP Creation instructions

s-rubenstein commented on docs/jade-getting-started.md:
````diff
@@ -145,14 +145,20 @@ helm repo update
 gcloud components install gke-gcloud-auth-plugin
 ```
 
+### 7. Configure
````
Should there be any text under `Configure`?

snf2ye replied:
Ah, thanks for catching this. I think I started a section here but ended up moving it.

s-rubenstein commented on docs/jade-getting-started.md:
```diff
@@ -311,30 +247,171 @@ psql -f db/create-data-repo-db
+* Open a new command line window, while bootRun runs in the background
+* In the new window, run `source ./render-configs.sh  -a integration` to pull secrets from vault and set environment variables in this context
+* Set this environment variable: `export IT_JADE_API_URL=http://localhost:8080`
+* Run `./gradlew :testIntegration --tests '*testSnapshotBuilder'` to run a specific integration test
```
Should `testSnapshotBuilder` instead be `testName` or something like that? (Or include `*testSnapshotBuilder` as an example later?)

s-rubenstein commented on docs/jade-getting-started.md:
````diff
@@ -311,30 +247,171 @@ psql -f db/create-data-repo-db
+```
+export PROXY_URL=http://localhost:8080
+```
+You need to have data repo running with `./gradlew bootRun` and the UI running with `npm start`.
````
Do the UI instructions make sense to be here, or should they be in `jade-data-repo-ui`? I would maybe advocate that the UI instructions should be stored primarily in the UI repo, and then there should be a pointer to them from here if we really want it.

snf2ye replied:
I think this is just supplemental instructions to that in the UI and there is a link directing the user to the UI readme. And, it's the piece that's relevant to setting up jade-data-repo for running with the UI.

s-rubenstein commented on render-configs.sh:
```diff
@@ -1,32 +1,161 @@
 #!/bin/bash
+### How to call this script
```
TOL (not required change): At some point we may want to think about converting this to python - I think it might give us better access to help text and a bit more help for function composition.

s-rubenstein commented on src/main/resources/application.properties:
```diff
@@ -141,7 +141,7 @@ azure.retryTimeoutSeconds=3600
 # ...
 
-rbs.enabled=false
+rbs.enabled=true
```
What is this change for?

snf2ye replied:
This is more of a drive-by: When I was looking at the env variables to set for RBS, I saw this was disabled by default. But, this really should not be the case as we need RBS any GCP setup. But, this flag is only used in our status endpoint - determining whether or not to report an RBS outage as a critical system. 

I'll just flip it back since it's not really relevant and I don't think it hurts anything to leave it as false, as we're overriding this in all of our production systems (https://github.com/broadinstitute/terra-helmfile/blob/c8fa6bb2f1ceedb4df70c6fdff8aa6110971e0cc/values/app/datarepo/live/prod.yaml#L108).

snf2ye committed:
Set the Azure B2C environment variables as well in render configs since they're required to set up auth in swagger

snf2ye committed:
Remove blank header and update numbering

snf2ye commented on docs/jade-getting-started.md:
```diff
@@ -312,29 +246,170 @@ psql --list
 
 Both are performed by a teammate in the Azure portal: https://portal.azure.com
 
+### 2. Create your own managed application in Azure
+You must have your own managed application in order to create a TDR azure billing profile.
+Create a "tdr-dev" managed application:
+* Azure portal -> Marketplace -> "My Marketplace" -> "Private plans" -> There you should see the "tdr-dev" plan.
+* Create a new tdr-dev plan with the following setup:
+  * Subscription: 8201558_TDR_testuser1 (if you don't have access, ask for help from team)
+  * Resource group: TDR
+  * Application Name: <your name>
+  * Hit "next"
+* On the next screen, ***pay attention to the email you set in this field. It will be the email
+you must log in as in order to create a TDR billing profile***. It should be a gmail account.
+* Hit create!
+
```
For reviewers: this covers [DC-921 -Write instructions for getting on Azure on TDR - setting up managed azure app](https://broadworkbench.atlassian.net/browse/DC-921). There are also instructions in the common problems doc with screen shots.

snf2ye commented on docs/jade-getting-started.md:
```diff
@@ -312,29 +246,170 @@ psql --list
+your bash profile (you may need to then restart intellij) or in the intellij run configuration for the selected test.
+* Select test in intellij UI, select 'testConnected' and run or debug it
+
+### 3. Run Integration tests
+We don't recommend running the entire integrated test suite locally, as it takes over two hours to run.
+Instead, you can select a specific test to run either in Intellij or the command line.
+First, make sure you have run through the following steps:
+* Start postgres
+* Ensure docker is running
+* You may need to re-auth with vault every so often. Run `vault login -method=github token=$(cat ~/.gh_token)`
+
+** Run test in the Command Line **
+* Run `source ./render-configs.sh -a integration` to pull secrets from vault, set environment variables and point to Integration Synapse on Azure
+* Start the app locally with `./gradlew bootRun`
+* Open a new command line window, while bootRun runs in the background
+* In the new window, run `source ./render-configs.sh  -a integration` to pull secrets from vault and set environment variables in this context
+* Set this environment variable: `export IT_JADE_API_URL=http://localhost:8080`
+* Run `./gradlew :testIntegration --tests '*testSnapshotBuilder'` to run a specific integration test
+
+** Run or Debug test in Intellij **
+* Run `./render-configs.sh  -a integration -p` to pull secrets from vault and print environment variables
+* Set the environments printed out from the render configs script. You can do this either in
+  your bash profile (you may need to then restart intellij) or in the intellij run configuration for the selected test.
+  * Additionally, set these environment variables:
+  * `export IT_JADE_API_URL=http://localhost:8080`
+* Start application by running `./gradlew bootRun`
+* Select test in intellij UI, select 'testIntegration' and run or debug it
```
Note for reviewers: This covers [DC-932 - Write instructions for getting on Azure on TDR - Running Azure integration tests locally](https://broadworkbench.atlassian.net/browse/DC-932).

snf2ye committed:
remove drive-by change

snf2ye commented on tools/setupResourceScripts/README.md:
```diff
@@ -31,3 +31,42 @@ in the "integration 4" environment using a new billing profile.
 "schema" is the name of the new directory and the "upload_prefix" is the gs path to the metadata files.
+
+
+### Azure OMOP Dataset and Snapshot Builder Settings Setup
+1. Prepare resources for creating a new azure billing profile (alternatively pass in an existing billing profile id with the --azure_profile_id flag)
+   * If you do not already have a managed application created, follow instructions in jade-getting-started.md
+   * Make note of the email account that set during managed app setup, you'll need to authenticate with this user
+   * Make note of the "Application name" set during managed app setup. You'll pass this in with the --azure_application_name flag
+   * Set the --billing_profile_file_name flag to the dev azure file: `--billing_profile_file_name='./files/azure_billing_profile_dev.json'`
+2. Determine TDR instance you want to populate (i.e.local, BEE, etc.)
+   * If running locally, set the --host flag to `http://localhost:8080`. Get TDR instance running locally by following instructions in jade-getting-started.md.
+   * Otherwise, set the --host flag to the appropriate TDR instance.
+3. If needed, set up your python instance:
+   * `python3 -m venv c:/path/to/myenv`
+   * `source c:/path/to/myenv/bin/activate`
+4. Authenticate with the email account you set up during managed app setup:
+   * `gcloud auth login <email>`
+5. Run the python script
+   * `cd jade-data-repo/tools/setupResourceScripts`
+   * `pip3 install -r requirements.txt`
+   * Run the set up script including all of the arguments, plus pointing to the azure OMOP dataset file:
+   * `python3 setup_tdr_resources.py --host <datarepo_url> --datasets='./files/azure_omop_dataset.json' --azure_application_name <azure_application_name> --billing_profile_file_name './files/azure_billing_profile_dev.json'`
+   * If you have a billing account already created, you can use the `--azure_profile_id` flag to pass in the billing profile id, then you no longer need to pass in some of the other flags:
+   * `python3 setup_tdr_resources.py --host <datarepo_url> --datasets='./files/azure_omop_dataset.json' --azure_profile_id <profile_id>`
+
+### GCP OMOP Dataset and Snapshot Builder Settings Setup
+1. Determine TDR instance you want to populate (i.e.local, BEE, etc.)
+  * If running locally, set the --host flag to `http://localhost:8080`. Get TDR instance running locally by following instructions in jade-getting-started.md.
+  * Otherwise, set the --host flag to the appropriate TDR instance.
+2. If needed, set up your python instance:
+  * `python3 -m venv c:/path/to/myenv`
+  * `source c:/path/to/myenv/bin/activate`
+3. Authenticate with the email account that has access to the GCP billing account `00708C-45D19D-27AAFA`:
+  * `gcloud auth login <email>`
+4. Run the python script
+  * `cd jade-data-repo/tools/setupResourceScripts`
+  * `pip3 install -r requirements.txt`
+  * Run the setup script including all the arguments, plus pointing to the GCP OMOP dataset file:
+  * `python3 setup_tdr_resources.py --host <datarepo_url> --datasets='./files/gcp_omop_dataset.json'`
+  * If you have a billing account already created, you can use the `--gcp_profile_id` flag to pass in the billing profile id.
```
Note to reviewers: These instructions cover [DC-930 - Write instructions for getting on Azure on TDR - setting up local OMOP DB](https://broadworkbench.atlassian.net/browse/DC-930)

snf2ye committed:
generic test name

snf2ye committed:
Remove defaults; move profiles to own folder; add setup instructions for UI

snf2ye commented on tools/setupResourceScripts/profiles/azure_billing_profile_dev.json:
@s-rubenstein  - I added a call in the help instructions to print out the available files for the billing profile files so I moved these to their own files. Let me know if you think this is an issue!

snf2ye committed:
shellcheck; print variables

s-rubenstein commented on tools/setupResourceScripts/datarepo_datasets_outputs.json:
```diff
@@ -1 +1 @@
-[{"V2F_GWAS_Summary_Statistics": {"dataset_id": "509ecb17-d442-420e-b9de-2870c858d931", "snapshot_ids": ["4f64f8b0-efa2-49bc-a509-7b7b27cdfe33"]}}]
```
We probably shouldn't check in the outputs - would you mind adding it to gitignore?

s-rubenstein reviewed (approved):
LGTM

snf2ye committed:
Rework to continue using z-shell for setting environment variables

typo/formatting

snf2ye committed:
Remove outputs, put in gitignore and move to outputs folder

snf2ye committed:
update gitignore and output path

snf2ye committed:
We need azure B2C settings for running locally, but not for tests

snf2ye commented on docs/jade-getting-started.md:
```diff
@@ -312,49 +246,65 @@ psql --list
 export GOOGLE_APPLICATION_CREDENTIALS=/tmp/jade-dev-account.json
 export GOOGLE_SA_CERT=/tmp/jade-dev-account.pem
 
-# Clears database on startup, test run, etc. This is further explained in the oncall playbook.
-export DB_MIGRATE_DROPALLONSTART=true
```
This causes trouble for our embedded datasets in unit tests. It has been our workflow to manually run `./gradlew dropall` in order to clear out our database instead.

snf2ye replied:
It actually appears to be an issue with the `TERRA_COMMON_STAIRWAY_FORCECLEANSTART` variable getting set to true. (It need to be set to false for connected tests to pass). 

But, I still don't know that I recommend the drop all on start variable. We're now seeding our local environments with datasets that we don't want to automatically get deleted.

snf2ye commented on docs/jade-getting-started.md:
```diff
@@ -363,43 +313,89 @@ export AZURE_SYNAPSE_SQLADMINPASSWORD=$(cat /tmp/jade-dev-synapse-admin-password
 export PACT_BROKER_USERNAME=$(cat /tmp/pact-ro-username.key)
 export PACT_BROKER_PASSWORD=$(cat /tmp/pact-ro-password.key)
 
-# Azure B2C authentication settings
-export OIDC_ADDCLIENTIDTOSCOPE=true
-export OIDC_AUTHORITYENDPOINT="https://oauth-proxy.dsp-eng-tools.broadinstitute.org/b2c"
-export OIDC_CLIENTID=bbd07d43-01cb-4b69-8fd0-5746d9a5c9fe
-export OIDC_EXTRAAUTHPARAMS="prompt=login"
-export OIDC_PROFILEPARAM=b2c_1a_signup_signin_tdr_dev
```
These settings cause our tests to fail. I've moved them to application.properties and overwritten them in our test properties to be empty.

snf2ye replied:
I haven't been able to reproduce this error. But, I've still moved them to application properties so that we don't need to manually set them.

snf2ye committed:
Add auth command needed for integration test setup

plus another update/typo

evansuslovich commented on tools/setupResourceScripts/README.md:
```diff
@@ -20,14 +20,62 @@ The script outputs are written to a JSON file in the format:
+  * `cd jade-data-repo/tools/setupResourceScripts`
+  * `pip3 install -r requirements.txt`
+  * Run the setup script including all the arguments, plus pointing to the GCP OMOP dataset file:
+  * `python3 setup_tdr_resources.py --host=<datarepo_url> --datasets='datarepo_gcp_omop_datasets.json'`
```
I think this should be changed to `'./suites/datarepo_gcp_omop_datasets.json'`

snf2ye replied:
Good eye! This is actually a change I'm making in this PR - I switched the code to look directly in the suites directory. This also makes it little more straight forward to read out the available files in the suites directory with the --help command. 

Hopefully this isn't too much of change to adjust to.

evansuslovich replied:
Oh, I see. In your script the file is appended to `./suites/`. Sweet!!!!

evansuslovich commented on tools/setupResourceScripts/README.md:
```diff
@@ -20,14 +20,62 @@ The script outputs are written to a JSON file in the format:
+* `python3 setup_tdr_resources.py --host=https://jade-4.datarepo-integration.broadinstitute.org --datasets='datarepo_azure_datasets.json' --billing_profile_file_name='azure_billing_profile_int.json'`
+* integration-5:
+* `python3 setup_tdr_resources.py --host=https://jade-5.datarepo-integration.broadinstitute.org --datasets='datarepo_datasets.json'`
+* `python3 setup_tdr_resources.py --host=https://jade-5.datarepo-integration.broadinstitute.org --datasets='datarepo_azure_datasets.json' --billing_profile_file_name='azure_billing_profile_int.json'`
```
I believe the datasets should also have a `./suites/`

evansuslovich replied:
I read your comment!

pshapiro4broad commented on render-configs.sh:
```diff
@@ -1,32 +1,136 @@
+echo $AZURE_SYNAPSE_WORKSPACENAME > "/tmp/azure-synapse-workspacename.txt"
+
+
+vault read -field=tenant-id secret/dsde/datarepo/"$AZURE_ENV"/azure-application-secrets \
+    > "/tmp/jade-dev-tenant-id.key"
+AZURE_CREDENTIALS_HOMETENANTID=$(cat "/tmp/jade-dev-tenant-id.key")
```
Not sure how much tweaking you want to do with this now, but since all the azure secrets are from the same vault path you can instead read them all at once and then pull out the fields you want from your local copy.

```
AZURE_SECRETS=$(vault read -field data -format json secret/dsde/datarepo/"$AZURE_ENV"/azure-application-secrets)

AZURE_CREDENTIALS_HOMETENANTID=$(echo "$AZURE_SECRETS" | jq -r '."tenant-id"')
AZURE_CREDENTIALS_APPLICATION_ID=$(echo "$AZURE_SECRETS" | jq -r '."client-id"')
...
```

snf2ye replied:
Thanks for taking a close look at this! 
It would be ideal to only need to need to pull the secret once, but I don't know that I find this echo/jq pattern much cleaner. And I still do need to write to the tmp files for our environment variables. I think I'm inclined to leave this piece as is.

pshapiro4broad commented on render-configs.sh:
```diff
@@ -1,32 +1,136 @@
-vault read -field=synapse-encryption-key secret/dsde/datarepo/dev/azure-application-secrets \
-    > /tmp/jade-dev-synapse-encryption-key.key
+if [[ "${COPY_INTELLIJ_ENV_VARS}" == "y" ]]; then
+  echo "AZURE_SYNAPSE_WORKSPACENAME=$AZURE_SYNAPSE_WORKSPACENAME;AZURE_CREDENTIALS_HOMETENANTID=$AZURE_CREDENTIALS_HOMETENANTID;AZURE_CREDENTIALS_APPLICATIONID=$AZURE_CREDENTIALS_APPLICATIONID;AZURE_CREDENTIALS_SECRET=$AZURE_CREDENTIALS_SECRET;AZURE_SYNAPSE_SQLADMINUSER=$AZURE_SYNAPSE_SQLADMINUSER;AZURE_SYNAPSE_SQLADMINPASSWORD=$AZURE_SYNAPSE_SQLADMINPASSWORD;AZURE_SYNAPSE_ENCRYPTIONKEY=$AZURE_SY ...
```
This could be done using a loop, which might be better, something like
```
VARIABLE_NAMES=(AZURE_SYNAPSE_WORKSPACENAME AZURE_CREDENTIALS_HOMETENANTID AZURE_CREDENTIALS_APPLICATIONID AZURE_CREDENTIALS_SECRET AZURE_SYNAPSE_SQLADMINUSER AZURE_SYNAPSE_SQLADMINPASSWORD" AZURE_SYNAPSE_ENCRYPTIONKEY GOOGLE_APPLICATION_CREDENTIALS GOOGLE_SA_CERT RBS_POOLID RBS_INSTANCEURL)

# Initialize an empty string
SETTINGS=""

# Loop over the array
for VAR_NAME in "${VARIABLE_NAMES[@]}"
do
  # Append the variable name and its value to the string
  SETTINGS+="$VAR_NAME=${!VAR_NAME};"
done

# Print the string
echo $SETTINGS
```

snf2ye replied:
That is better, thanks!

pshapiro4broad commented on render-configs.sh:
```diff
@@ -1,32 +1,136 @@
+# ========================
+# Google Credentials
+# ========================
 vault read -format=json secret/dsde/datarepo/dev/sa-key.json \
     | jq .data | tee /tmp/jade-dev-account.json \
     | jq -r .private_key > /tmp/jade-dev-account.pem
```
You could avoid a call to `jq` here
```suggestion
vault read -field data -format=json secret/dsde/datarepo/dev/sa-key.json \
    | tee /tmp/jade-dev-account.json \
    | jq -r .private_key > /tmp/jade-dev-account.pem
```

s-rubenstein commented on src/main/resources/application.properties:
```diff
@@ -49,12 +49,12 @@ oauth.loginEndpoint=https://accounts.google.com/o/oauth2/auth
 # where configuration varies from one environment to the next
 oidc.schemeName=oidc
-oidc.clientId=
+oidc.clientId=bbd07d43-01cb-4b69-8fd0-5746d9a5c9fe
```
Just to confirm - clientId is not secret, right?

snf2ye replied:
No it's not, but thanks for checking! It was also included in a [previous PR](https://github.com/DataBiosphere/jade-data-repo/pull/1624/files#diff-54477201e1f84732669676e3a829a4ff30c6d7b56655addffbb8f2fda49b982dR373)

s-rubenstein reviewed (approved):
LGTM pending the `~/.zshrc` change

snf2ye committed:
Fix order of /.

snf2ye committed:
test remove

snf2ye committed:
Add note

snf2ye committed:
Loop through variables

snf2ye committed:
code review: remove need to jq
